File: fashion-recommend/traits/gorse_sync.go
package traits

import (
	"encoding/json"
	"fmt"

	"fashion-recommend/client"
	"fashion-recommend/database"
	"fashion-recommend/models"
)

// GorseSync Gorse 同步服务
type GorseSync struct {
	gorseClient *client.GorseClient
	db          *database.DB
}

// NewGorseSync 创建 Gorse 同步服务
func NewGorseSync(gorseClient *client.GorseClient, db *database.DB) *GorseSync {
	return &GorseSync{
		gorseClient: gorseClient,
		db:          db,
	}
}

// SyncUserTraitsToGorse 同步用户特质到 Gorse
func (g *GorseSync) SyncUserTraitsToGorse(userID string) error {
	// 1. 从数据库获取用户特质
	userTraits, err := g.db.GetUserTraits(userID)
	if err != nil {
		return fmt.Errorf("获取用户特质失败: %w", err)
	}

	if userTraits == nil {
		// 用户还没有特质数据
		return nil
	}

	// 2. 解析特质数据
	var traits database.TraitsData
	if err := json.Unmarshal(userTraits.Traits, &traits); err != nil {
		return fmt.Errorf("解析特质数据失败: %w", err)
	}

	// 3. 转换为 Gorse Labels
	labels := g.convertTraitsToLabels(&traits)

	// 4. 更新 Gorse 用户
	user := models.User{
		UserId:  userID,
		Labels:  labels,
		Comment: fmt.Sprintf("特质置信度: %.2f", userTraits.ConfidenceScore),
	}

	if err := g.gorseClient.InsertUser(user); err != nil {
		return fmt.Errorf("更新 Gorse 用户失败: %w", err)
	}

	return nil
}

// convertTraitsToLabels 将特质转换为 Gorse Labels
func (g *GorseSync) convertTraitsToLabels(traits *database.TraitsData) []string {
	labels := []string{}

	// 风格偏好（只添加分数 > 0.5 的）
	for style, score := range traits.StylePreferences {
		if score > 0.5 {
			labels = append(labels, fmt.Sprintf("style:%s:%.1f", style, score))
		}
	}

	// 颜色偏好（只添加分数 > 0.5 的）
	for color, score := range traits.ColorPreferences {
		if score > 0.5 {
			labels = append(labels, fmt.Sprintf("color:%s:%.1f", color, score))
		}
	}

	// 价格敏感度
	if traits.PriceSensitivity != "" {
		labels = append(labels, fmt.Sprintf("price:%s", traits.PriceSensitivity))
	}

	// 品牌偏好
	for _, brand := range traits.BrandPreferences {
		labels = append(labels, fmt.Sprintf("brand:%s", brand))
	}

	// 使用场景
	for _, occasion := range traits.Occasions {
		labels = append(labels, fmt.Sprintf("occasion:%s", occasion))
	}

	// 兴趣标签
	for _, interest := range traits.Interests {
		labels = append(labels, fmt.Sprintf("interest:%s", interest))
	}

	// 关键词（最多取前 5 个）
	keywordCount := len(traits.Keywords)
	if keywordCount > 5 {
		keywordCount = 5
	}
	for i := 0; i < keywordCount; i++ {
		labels = append(labels, fmt.Sprintf("keyword:%s", traits.Keywords[i]))
	}

	return labels
}

// GenerateImplicitFeedback 根据对话生成隐式反馈
func (g *GorseSync) GenerateImplicitFeedback(userID string, traits *database.TraitsData) error {
	// 这里可以根据用户特质生成一些隐式的"喜欢"反馈
	// 例如：如果用户喜欢简约风格，可以给简约风格的商品添加隐式反馈
	
	// 暂时不实现，留作后续优化
	return nil
}

File: python-agent/traits/gorse_sync.py
"""Gorse trait-label sync — port of fashion-recommend/traits/gorse_sync.go.

SyncUserTraitsToGorse() → sync_user_traits():
  1. Fetch user_traits from DB.
  2. Convert TraitsData fields to Gorse label strings.
  3. POST the updated user to Gorse (/api/user).

Label format is identical to the Go implementation:
  style:{name}:{score:.1f}
  color:{name}:{score:.1f}
  price:{sensitivity}
  brand:{name}
  occasion:{name}
  interest:{name}
  keyword:{kw}   (top 5 only)
"""

from __future__ import annotations

from db.client import DBClient
from db.gorse_client import GorseClient


class GorseSync:
    def __init__(self, db: DBClient, gorse: GorseClient) -> None:
        self._db = db
        self._gorse = gorse

    async def sync_user_traits(self, user_id: str) -> None:
        """Fetch traits from DB and push label-encoded version to Gorse."""
        row = await self._db.get_user_traits(user_id)
        if row is None:
            return

        traits: dict = row.get("traits") or {}
        labels = self._traits_to_labels(traits)
        confidence: float = row.get("confidence_score", 0.0)
        comment = f"特质置信度: {confidence:.2f}"

        await self._gorse.insert_user(user_id, labels, comment)

    @staticmethod
    def _traits_to_labels(traits: dict) -> list[str]:
        labels: list[str] = []

        for style, score in (traits.get("style_preferences") or {}).items():
            if score > 0.5:
                labels.append(f"style:{style}:{score:.1f}")

        for color, score in (traits.get("color_preferences") or {}).items():
            if score > 0.5:
                labels.append(f"color:{color}:{score:.1f}")

        price = traits.get("price_sensitivity", "")
        if price:
            labels.append(f"price:{price}")

        for brand in (traits.get("brand_preferences") or []):
            labels.append(f"brand:{brand}")

        for occasion in (traits.get("occasions") or []):
            labels.append(f"occasion:{occasion}")

        for interest in (traits.get("interests") or []):
            labels.append(f"interest:{interest}")

        for kw in (traits.get("keywords") or [])[:5]:
            labels.append(f"keyword:{kw}")

        return labels
